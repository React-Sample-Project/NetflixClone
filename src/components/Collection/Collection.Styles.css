.h-collectioncustomization {
    margin-bottom: 4vw;
}

.h-collectioncustomization.slick-slider {
    padding: 0 4%;
}

.h-collectioncustomization .slick-list {   
    padding: 0 !important;
    overflow: unset;

}

@media screen and (min-width: 1500px) {
    .h-collectioncustomization.slick-slider {
        padding: 0 60px;
    }
}