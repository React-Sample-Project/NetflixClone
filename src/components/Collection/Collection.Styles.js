import styled from "styled-components";

export const Header = styled.div`
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  padding-bottom: 25px;
  margin-left: 60px;
`;

export const HeaderText = styled.span`
  font-size: 30px;
  font-weight: 700;
  white-space: nowrap;
  margin-right: 15px;
`;
