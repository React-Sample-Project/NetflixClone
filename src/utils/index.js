export const generateLoadingItems = () => [
  {
    genreId: Math.floor(Math.random() * 1000000000),
    isLoading: true,
  },
];

export const isBottom = (element) =>
  element?.getBoundingClientRect().bottom <= window.innerHeight + 400;

export const isContentOverflown = ({
  clientWidth,
  clientHeight,
  scrollWidth,
  scrollHeight,
}) => scrollHeight > clientHeight || scrollWidth > clientWidth;

export const chunkArrays = (array, chunkLength) => {
  const arrayLength = array.length;
  let result = [];
  const currentLength = 0;
  let actualChunkLength = Math.min(array.length, chunkLength);
  for (let i = currentLength; i < arrayLength; i += chunkLength) {
    result.push(array.slice(i, i + actualChunkLength));
  }
  return result;
};

export const isArraysEqual = (a, b) => {
  if (a == null || b == null) return false;
  if (a.length !== b.length) return false;
  for (var i = 0; i < a.length; ++i) {
    if (a[i] !== b[i]) return false;
  }
  return true;
};

export const debounce = (func, wait) => {
  let timeout;
  return function (...args) {
    const context = this;
    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(() => {
      timeout = null;
      func.apply(context, args);
    }, wait);
  };
};

export const convertToHours = (mins) => {
  if (mins) {
    const minutes = mins % 60;
    const hrs = (mins - minutes) / 60;
    return hrs + "h " + minutes + "m";
  }
};

export const getStringWithComma = (array) =>
  array.map(({ name }) => name).join(",");

export const getCorrectedMediaType = (mediaType) =>
  mediaType.includes("movie") ? "movie" : mediaType;

export const getNumArray = (length = 6) => [...Array(length).keys()];
