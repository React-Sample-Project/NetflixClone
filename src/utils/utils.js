export const generateLoadingItems = () => [
  {
    genreId: Math.floor(Math.random() * 1000000000),
    isLoading: true,
  },
];

export const isBottom = (element) =>
  element && element.getBoundingClientRect().bottom <= window.innerHeight + 40;

export const isContentOverflown = ({
  clientWidth,
  clientHeight,
  scrollWidth,
  scrollHeight,
}) => scrollHeight > clientHeight || scrollWidth > clientWidth;

export const chunkArrays = (array, chunkLength) => {
  const arrayLength = array.length;
  let result = [];
  const currentLength = 0;
  let actualChunkLength = Math.min(array.length, chunkLength);
  for (let i = currentLength; i < arrayLength; i += chunkLength) {
    result.push(array.slice(i, i + actualChunkLength));
  }
  // console.log(result);
  return result;
};

export const debounce = (func, wait) => {
  let timeout;
  return function (...args) {
    const context = this;
    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(() => {
      timeout = null;
      func.apply(context, args);
    }, wait);
  };
};
