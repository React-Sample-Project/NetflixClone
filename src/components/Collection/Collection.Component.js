import React, { useEffect, useState, useRef } from "react";
import CollectionSlider from "../CollectionSlider/CollectionSlider.Component";
import useScroll from "../../hooks/useScroll";
import useFetch from "../../hooks/useFetch";

import { chunkArrays, isBottom } from "../../utils/utils";
import "./Collection.Styles.css";
import DataFetchConstants from "../../store/DataFetch/DataFetch.Constants";

function Collection({ fetchMethod, args }) {
  const [slicedCollection, setSlicedCollection] = useState([]);
  const rootRef = useRef(null);
  //To do: for responsiveness
  const length = 6;
  //Math.ceil(document.documentElement.clientWidth / 318)
  const [{ data: collection, currentPage, totalPages }, dispatch] = useFetch(
    fetchMethod,
    null,
    args,
    {
      paging: true,
    }
  );
  useScroll(() => {
    if (isBottom(rootRef.current) && currentPage <= totalPages) {
      dispatch({ type: DataFetchConstants.INCREMENT_CURRENT_PAGE });
    }
  });

  useEffect(() => {
    dispatch({
      type: DataFetchConstants.RESET_STATE,
      payload: {
        paging: true,
        data: null,
      },
    });
  }, [dispatch, args]);

  useEffect(() => {
    if (collection) {
      const collectionLength = collection.length;
      if (collectionLength > 0) {
        //  const mod = collectionLength % length;
        //let finalArray = collection;
        // if (mod !== 0 && mod !== 1) {
        // !!!important To do: Fix this. Not working when there is only one page and for example if there are 11 elements. Then only first 6 is displayed and the rest is ignored.
        // finalArray = collection.slice(0, collectionLength - mod);
        // }
        setSlicedCollection(chunkArrays(collection, length));
      }
    }
  }, [collection]);
  return (
    <div style={{ paddingTop: "3%", position: "relative" }} ref={rootRef}>
      {slicedCollection.map((collection) => (
        <CollectionSlider
          className="genre-customization"
          items={collection}
          isLoading={collection[collection.length - 1].isLoading}
          key={collection[0].id}
          // length={Math.min(length, collection.length)}
          doNotAddControls={true}
        />
      ))}
    </div>
  );
}

export default Collection;
