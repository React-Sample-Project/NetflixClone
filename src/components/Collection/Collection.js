import React, { useEffect, useState, useRef } from "react";
import CollectionSlider from "../CollectionSlider/CollectionSlider";
import useScroll from "../../hooks/useScroll";
import useFetch from "../../hooks/useFetch";

import { chunkArrays, isBottom } from "../../utils";
import "./Collection.Styles.css";
import { DataFetchConstants } from "../../store/reducers/AsyncFetchArray";
import { Header, HeaderText } from "./Collection.Styles";

function Collection({ fetchMethod, args, title, onResultsChange }) {
  const [slicedCollection, setSlicedCollection] = useState([]);
  const rootRef = useRef(null);
  // To do: for responsiveness
  const length = 6;
  // Math.ceil(document.documentElement.clientWidth / 318)
  const [
    { data: collection, currentPage, totalPages },
    dispatch,
    asyncFunction,
  ] = useFetch(fetchMethod, null, {
    paging: true,
  });

  useScroll(() => {
    if (isBottom(rootRef.current) && currentPage < totalPages) {
      dispatch({ type: DataFetchConstants.INCREMENT_CURRENT_PAGE });
    }
  });

  useEffect(() => {
    onResultsChange && onResultsChange(collection);
  }, [onResultsChange, collection]);

  useEffect(() => {
    // To prevent calling the same function twice at first time when arguments change, this condition is added
    if (currentPage > 1) {
      asyncFunction(...args, currentPage);
    }
    // args is not included in the dependency array since separate useEffect exists for args change and args and currentPage will not change at the same time.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [asyncFunction, currentPage]);

  useEffect(() => {
    dispatch({
      type: DataFetchConstants.RESET_STATE,
      payload: {
        data: null,
      },
    });
    asyncFunction(...args);
  }, [dispatch, asyncFunction, args]);

  useEffect(() => {
    if (collection) {
      setSlicedCollection(chunkArrays(collection, length));
    }
  }, [collection]);
  return (
    <div style={{ paddingTop: "1%", position: "relative" }} ref={rootRef}>
      <Header>
        <HeaderText>{title}</HeaderText>
      </Header>
      {slicedCollection.map((collection) => (
        <CollectionSlider
          className="collection-customization"
          items={collection}
          isLoading={collection[collection.length - 1].isLoading}
          key={collection[0].id || collection[0].genreId}
          // length={Math.min(length, collection.length)}
          doNotAddControls={true}
        />
      ))}
    </div>
  );
}

export default Collection;
