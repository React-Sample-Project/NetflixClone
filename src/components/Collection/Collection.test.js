import React from "react";
import { render } from "@testing-library/react";
import Collection from "./Collection";
import { generateLoadingItems } from "../../utils";

describe("Testing Collection Component", () => {
  const asyncFunc = jest.fn(() =>
    Promise.resolve({ data: [...generateLoadingItems()] })
  );
  it("displays loading skeleton", () => {
    const { container } = render(
      <Collection fetchMethod={asyncFunc} args={[1, 2]} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
