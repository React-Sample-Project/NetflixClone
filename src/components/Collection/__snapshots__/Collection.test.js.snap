// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Collection Component displays loading skeleton 1`] = `
<div
  style="padding-top: 1%; position: relative;"
>
  <div
    class="sc-fzoLag jndESe"
  >
    <span
      class="sc-fzoXzr jnpFzy"
    />
  </div>
  <div
    class="slick-slider slider__inner h-collectioncustomization slick-initialized"
    dir="ltr"
  >
    <div
      class="slick-list"
      style="height: 0px;"
    >
      <div
        class="slick-track"
        style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
      >
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="-4"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="-3"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="-2"
          style="width: 0px;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="-1"
          style="width: 0px;"
          tabindex="-1"
        />
        <div
          aria-hidden="false"
          class="slick-slide slick-active slick-current"
          data-index="0"
          style="outline: none; width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="false"
          class="slick-slide slick-active"
          data-index="1"
          style="outline: none; width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="false"
          class="slick-slide slick-active"
          data-index="2"
          style="outline: none; width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="false"
          class="slick-slide slick-active"
          data-index="3"
          style="outline: none; width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide"
          data-index="4"
          style="outline: none; width: 0px;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          class="slick-slide"
          data-index="5"
          style="outline: none; width: 0px;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="6"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="7"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="8"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="9"
          style="width: 0px;"
          tabindex="-1"
        >
          <div>
            <div
              class="sc-fzqNJr mClsx"
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="10"
          style="width: 0px;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          class="slick-slide slick-cloned"
          data-index="11"
          style="width: 0px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</div>
`;
